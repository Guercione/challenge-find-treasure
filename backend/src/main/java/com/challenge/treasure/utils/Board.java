package com.challenge.treasure.utils;

import java.util.ArrayList;

public class Board {
    public ArrayList<ArrayList<Integer>> generateInitialBoard() {
        ArrayList<ArrayList<Integer>> board = new ArrayList<ArrayList<Integer>>();

        for(int x = 0; x < 5; x++) {
            board.add(new ArrayList<Integer>());
            for (int y = 0; y < 5; y++) {
                board.get(x).add(0);
            }
        }

        return board;
    }

    public ArrayList<ArrayList<Integer>> generateCompleteBoard() {
        ArrayList<ArrayList<Integer>> board = new ArrayList<ArrayList<Integer>>();

        for(int x = 0; x < 5; x++) {
            board.add(new ArrayList<Integer>());
        }

        board.get(0).add(3);
        board.get(0).add(9);
        board.get(0).add(3);
        board.get(0).add(2);
        board.get(0).add(1);

        board.get(1).add(2);
        board.get(1).add(3);
        board.get(1).add(2);
        board.get(1).add(3);
        board.get(1).add(2);

        board.get(2).add(1);
        board.get(2).add(2);
        board.get(2).add(3);
        board.get(2).add(9);
        board.get(2).add(3);

        board.get(3).add(2);
        board.get(3).add(3);
        board.get(3).add(2);
        board.get(3).add(3);
        board.get(3).add(2);

        board.get(4).add(3);
        board.get(4).add(9);
        board.get(4).add(3);
        board.get(4).add(2);
        board.get(4).add(1);

        return board;
    }
}
