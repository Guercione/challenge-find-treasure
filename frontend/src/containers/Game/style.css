.game-content {
  height: 100vh;
  display: flex;
  align-items: center;
  justify-content: space-around;
}

.game-personal-score {
  display: flex;
  flex-direction: column;
}

.game-start {
  width: 300px;
  display: flex;
  flex-direction: column;
}
