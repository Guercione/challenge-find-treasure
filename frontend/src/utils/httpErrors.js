export const requestStatusCheck = (api, action = "server") => {
  const httpErrors = {
    200: { data: api.data, status: api.status, error: undefined },
    400: {
      data: api.data,
      status: api.status,
      error: "Some fields are missing",
    },
    404: {
      data: api.data,
      status: api.status,
      error: "No match found for this user",
    },
    500: {
      data: api.data,
      status: api.status,
      error: `Error on ${action} try again later`,
    },
    999: {
      data: api?.data,
      status: api?.status,
      error: `Unknow error, try again later`,
    },
  };

  return httpErrors[api.status] || httpErrors[999];
};
