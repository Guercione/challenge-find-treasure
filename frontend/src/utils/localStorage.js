const _match = "_match";
const _userName = "_username";

export const setMatch = (matchObject) => {
  return localStorage.setItem(_match, JSON.stringify(matchObject) || {});
};

export const getMatch = () => {
  return JSON.parse(localStorage.getItem(_match) || "{}");
};

export function deletetMatch() {
  localStorage.removeItem(_match);
}

export const setUserName = (userName) => {
  return localStorage.setItem(_userName, userName);
};

export const getUserName = () => {
  return localStorage.getItem(_userName);
};
