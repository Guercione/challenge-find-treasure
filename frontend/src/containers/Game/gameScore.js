import React from "react";
import PropTypes from "prop-types";

import { connect } from "react-redux";
import { gamePostRestartSaga } from "redux/actions/gameAction";

import "./style.css";

const PersonalScore = ({ turns, gamePostRestartSaga }) => {
  return (
    <div className="game-personal-score">
      <span>Your score: {turns}</span>
      <button onClick={gamePostRestartSaga}>Play again</button>
    </div>
  );
};

PersonalScore.propTypes = {
  turns: PropTypes.number,
  gamePostRestartSaga: PropTypes.func.isRequired,
};

PersonalScore.defaultProps = {
  turns: 0,
};

export default connect(
  ({ game }) => ({
    turns: game.turns,
  }),
  { gamePostRestartSaga }
)(PersonalScore);
