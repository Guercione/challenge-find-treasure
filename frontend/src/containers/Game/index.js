import React from "react";
import PropTypes from "prop-types";

import { connect } from "react-redux";
import { gameGetGeneralScoreSaga } from "redux/actions/gameAction";

import Start from "./start";
import Board from "containers/Board";
import GameScore from "./gameScore";
import GeneralScore from "containers/GeneralScore";
import If from "components/If";
import Loading from "components/Loading";

import "./style.css";

const Game = ({
  userName,
  board,
  matchHash,
  done,
  loading,
  error,
  gameGetGeneralScoreSaga,
}) => {
  const renderContent = () => {
    if (done) return <GameScore />;

    if (userName && matchHash && board.length) {
      return <Board />;
    }

    return <Start />;
  };

  React.useEffect(() => {
    gameGetGeneralScoreSaga();
  }, [done]);

  return (
    <React.Fragment>
      <div className="game-content">
        <div>{renderContent()}</div>
        <div>
          <If condition={userName}>
            <div>Username: {userName}</div>
          </If>

          <h3>Legends:</h3>
          <p>T - Treasure</p>
          <p>3 - Closest proximity</p>
          <p>2 - Diagonal proximity, second to '3'</p>
          <p>1 - Second to '2'</p>

          <If condition={loading}>
            <Loading />
          </If>

          <If condition={error.active}>
            <div>
              {error.type}: {error.message}
            </div>
          </If>
        </div>
        <GeneralScore />
      </div>
    </React.Fragment>
  );
};

Game.propTypes = {
  done: PropTypes.bool,
  userName: PropTypes.string,
  matchHash: PropTypes.number,
  error: PropTypes.shape({
    type: PropTypes.string,
    active: PropTypes.bool,
    message: PropTypes.string,
  }),
  board: PropTypes.array.isRequired,
  loading: PropTypes.bool.isRequired,
  gameGetGeneralScoreSaga: PropTypes.func.isRequired,
};

Game.defaultProps = {
  done: false,
  userName: "",
  matchHash: 0,
  error: {
    type: "",
    active: false,
    message: "",
  },
};

export default connect(
  ({ game, notification }) => ({
    loading: game.loading,
    userName: game.userName,
    board: game.userBoard,
    matchHash: game.matchHash,
    done: game.done,
    error: notification,
  }),
  { gameGetGeneralScoreSaga }
)(Game);
