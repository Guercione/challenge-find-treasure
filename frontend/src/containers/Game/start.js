import React from "react";
import PropTypes from "prop-types";

import { connect } from "react-redux";
import { gamePostNewGameSaga } from "redux/actions/gameAction";

import "./style.css";
import If from "components/If";

const Start = ({ gamePostNewGameSaga }) => {
  const [name, setName] = React.useState("");
  const [error, setError] = React.useState("");

  function handlePlay() {
    if (name && name.length >= 4 && RegExp(/^[a-zA-Z0-9]*$/).test(name)) {
      gamePostNewGameSaga(name);
      setError("");
      return;
    }

    setError(
      "Name must contain at minimum 4 characters, letters and numbers only"
    );
  }

  return (
    <div className="game-start">
      <label>Insert your username:</label>
      <input
        type="text"
        value={name}
        onChange={(e) => setName(e.target.value)}
      />
      <button onClick={handlePlay}>Play</button>
      <If condition={error}>
        <span>{error}</span>
      </If>
    </div>
  );
};

Start.propTypes = {
  gamePostNewGameSaga: PropTypes.func.isRequired,
};

Start.defaultProps = {};

export default connect(null, { gamePostNewGameSaga })(Start);
