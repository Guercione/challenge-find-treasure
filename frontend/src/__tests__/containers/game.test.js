import React from "react";
import renderer from "react-test-renderer";

import { renderWithReduxAllReducers } from "utils/unitTest";

import Game from "containers/Game";

describe("CONTAINER - Game", () => {
  it("Game - Test of board render correctly", () => {
    const tree = renderer
      .create(renderWithReduxAllReducers(<Game />).container.innerHTML)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
