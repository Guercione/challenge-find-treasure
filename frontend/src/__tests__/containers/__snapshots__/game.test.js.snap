// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CONTAINER - Game Game - Test of board render correctly 1`] = `"<div class=\\"game-content\\" data-testid=\\"game-content\\"><div><div class=\\"game-start\\"><label>Insert your username:</label><input type=\\"text\\" value=\\"\\"><button>Play</button><div></div></div></div><div><div></div><h3>Legends:</h3><p>T - Treasure</p><p>3 - Closest proximity</p><p>2 - Diagonal proximity, second to '3'</p><p>1 - Second to '2'</p><div><svg svg=\\"http://www.w3.org/2000/svg\\" xmlns=\\"http://www.w3.org/2000/svg\\" link=\\"http://www.w3.org/1999/xlink\\" version=\\"1.0\\" width=\\"30px\\" height=\\"30px\\" viewBox=\\"0 0 128 128\\" pace=\\"preserve\\"><g><path d=\\"M75.4 126.63a11.43 11.43 0 0 1-2.1-22.65 40.9 40.9 0 0 0 30.5-30.6 11.4 11.4 0 1 1 22.27 4.87h.02a63.77 63.77 0 0 1-47.8 48.05v-.02a11.38 11.38 0 0 1-2.93.37z\\" fill=\\"#636b70\\" fill-opacity=\\"1\\"></path><animateTransform attributeName=\\"transform\\" type=\\"rotate\\" from=\\"0 64 64\\" to=\\"360 64 64\\" dur=\\"1000ms\\" repeatCount=\\"indefinite\\"></animateTransform></g></svg></div><div></div></div><div class=\\"generalscore-general-score\\"><h4>TOP 10 Best Players</h4><div>No score found</div><div></div></div></div>"`;
